esinstall/src/entrypoints.ts

```diff
@@ -0,0 +1,337 @@
+import {readdirSync, existsSync, realpathSync} from 'fs';
+import path from 'path';
+import validatePackageName from 'validate-npm-package-name';
+import {ExportField, ExportMapEntry, PackageManifestWithExports, PackageManifest} from './types';
+import {parsePackageImportSpecifier, resolveDependencyManifest} from './util';
+import pm from 'picomatch';
+
+// Rarely, a package will ship a broken "browser" package.json entrypoint.
+// Ignore the "browser" entrypoint in those packages.
+const BROKEN_BROWSER_ENTRYPOINT = ['@sheerun/mutationobserver-shim'];
+
+function hasTypes(manifest: PackageManifest): boolean {
+  return !!(manifest.types || manifest.typings);
+}
+
+type FindManifestEntryOptions = {
+  packageLookupFields?: string[];
+  packageName?: string;
+};
+
+/**
+ *
+ */
+export function findManifestEntry(
+  manifest: PackageManifest,
+  entry?: string,
+  {packageLookupFields = [], packageName}: FindManifestEntryOptions = {},
+): string | undefined {
+  let foundEntrypoint: string | undefined;
+
+  if (manifest.exports) {
+    foundEntrypoint =
+      typeof manifest.exports === 'string'
+        ? manifest.exports
+        : findExportMapEntry(manifest.exports['.']);
+
+    if (typeof foundEntrypoint === 'string') {
+      return foundEntrypoint;
+    }
+  }
+
+  foundEntrypoint = [
+    ...packageLookupFields,
+    'browser:module',
+    'module',
+    'main:esnext',
+    'jsnext:main',
+  ]
+    .map((e) => manifest[e])
+    .find(Boolean);
+
+  if (foundEntrypoint) {
+    return foundEntrypoint;
```

Comment:
Great! This falls back to the `package.json` entrypoints if it can’t find the correct export map entry, so this would solve the `preact@10.3.0` issue of having an export map but accidentally omitting `['.']` (which is technically invalid, but I think this is a better experience for users)